File: gear-commons/src/main/scala/cn/howardliu/gear/commons/utils/SystemInfo.scala
package cn.howardliu.gear.commons.utils

import java.io.{InputStream, InputStreamReader}
import java.lang.management.{ManagementFactory, OperatingSystemMXBean, RuntimeMXBean}
import java.lang.reflect.Method
import java.nio.charset.Charset
import java.text.{DecimalFormat, DecimalFormatSymbols}
import java.util.{Locale, Map => JMap}

import Type._
import org.apache.commons.io.IOUtils

import scala.collection.immutable.HashMap

/**
 * <br/>created at 16-3-21
 * @author liuxh
 * @since 1.1.11
 */
object SystemInfo {
  val ONE_KB: Double = 1024.toDouble
  val ONE_MB: Double = ONE_KB * ONE_KB
  val ONE_GB: Double = ONE_KB * ONE_MB

  def getSystemInfo: JMap[String, Any] = {
    val os: OperatingSystemMXBean = ManagementFactory.getOperatingSystemMXBean
    var info = HashMap("name" -> os.getName, "version" -> os.getVersion, "arch" -> os.getArch, "systemLoadAverage" -> os.getSystemLoadAverage)

    // com.sun.management.OperatingSystemMXBean
    info += ("committedVirtualMemorySize" -> addGetterIfAvaliable(os, "committedVirtualMemorySize"))
    info += ("freePhysicalMemorySize" -> addGetterIfAvaliable(os, "freePhysicalMemorySize"))
    info += ("freeSwapSpaceSize" -> addGetterIfAvaliable(os, "freeSwapSpaceSize"))
    info += ("processCpuTime" -> addGetterIfAvaliable(os, "processCpuTime"))
    info += ("totalPhysicalMemorySize" -> addGetterIfAvaliable(os, "totalPhysicalMemorySize"))
    info += ("totalSwapSpaceSize" -> addGetterIfAvaliable(os, "totalSwapSpaceSize"))

    // com.sun.management.UnixOperatingSystemMXBean
    info += ("openFileDescriptorCount" -> addGetterIfAvaliable(os, "openFileDescriptorCount"))
    info += ("maxFileDescriptorCount" -> addGetterIfAvaliable(os, "maxFileDescriptorCount"))

    try {
      if (!os.getName.toLowerCase(Locale.ROOT).startsWith("windows")) {
        info += ("uname" -> execute("uname -a"))
        info += ("uptime" -> execute("uptime"))
      }
    } catch {
      case _: Exception =>
    }
    >>(info)
  }

  def getJvmInfo: JMap[String, Any] = {
    val javaVersion: String = System.getProperty("java.specification.version", "unknown")
    val javaVendor: String = System.getProperty("java.specification.vendor", "unknown")
    val javaName: String = System.getProperty("java.specification.name", "unknown")
    val jreVersion: String = System.getProperty("java.version", "unknown")
    val jreVendor: String = System.getProperty("java.vendor", "unknown")
    val vmVersion: String = System.getProperty("java.vm.version", "unknown")
    val vmVendor: String = System.getProperty("java.vm.vendor", "unknown")
    val vmName: String = System.getProperty("java.vm.name", "unknown")

    var jvm: HashMap[String, Any] = HashMap("version" -> (jreVersion + " " + vmVersion), "name" -> (jreVendor + " " + vmName))
    jvm += ("spec" -> HashMap("vendor" -> javaVendor, "name" -> javaName, "version" -> javaVersion))
    jvm += ("jre" -> HashMap("vendor" -> jreVendor, "version" -> jreVersion))
    jvm += ("vm" -> HashMap("vendor" -> vmVendor, "name" -> vmName, "version" -> vmVersion))

    val runtime: Runtime = Runtime.getRuntime
    val free: Long = runtime.freeMemory
    val max: Long = runtime.maxMemory
    val total: Long = runtime.totalMemory
    val used: Long = total - free
    val percentUsed: Double = (used.toDouble / max.toDouble) * 100

    jvm += ("processors" -> runtime.availableProcessors)
    val df = new DecimalFormat("#.##", DecimalFormatSymbols.getInstance(Locale.ROOT))
    var mem: HashMap[String, Any] = HashMap("free" -> humanReadableUnits(free, df), "used" -> (humanReadableUnits(used, df) + " (%" + df.format(percentUsed) + ")"), "total" -> humanReadableUnits(total, df), "max" -> humanReadableUnits(max, df))
    mem += ("raw" -> HashMap("free" -> free, "used" -> used, "used%" -> percentUsed, "total" -> total, "max" -> max))
    jvm += ("memory" -> mem)

    val mx: RuntimeMXBean = ManagementFactory.getRuntimeMXBean
    jvm += ("jmx" -> HashMap("bootclasspath" -> mx.getBootClassPath, "classpath" -> mx.getClassPath, "commandLineArgs" -> mx.getInputArguments, "startTime" -> mx.getStartTime, "upTimeMS" -> mx.getUptime))
    >>(jvm)
  }

  private def humanReadableUnits(bytes: Long, df: DecimalFormat) = if ((bytes / ONE_GB).toLong > 0) df.format(bytes / ONE_GB) + "GB" else if ((bytes / ONE_MB).toLong > 0) df.format(bytes / ONE_MB) + "MB" else if ((bytes / ONE_KB).toLong > 0) df.format(bytes / ONE_KB) + "KB" else df.format(bytes) + "bytes"

  private def addGetterIfAvaliable(obj: AnyRef, getter: String): Any = {
    try {
      val n: String = Character.toUpperCase(getter.charAt(0)) + getter.substring(1)
      val m: Method = obj.getClass.getMethod("get" + n)
      m.setAccessible(true)
      m.invoke(obj, null.asInstanceOf[Array[AnyRef]])
    } catch {
      case ignored: Exception => ""
    }
  }

  private def execute(cmd: String): String = {
    var in: InputStream = null
    var process: Process = null
    try {
      process = Runtime.getRuntime.exec(cmd)
      in = process.getInputStream
      IOUtils.toString(new InputStreamReader(in, Charset.defaultCharset))
    } catch {
      case ex: Exception => "(error executing: " + cmd + ")"
      case err: Error =>
        if (err.getMessage != null && (err.getMessage.contains("posix_spawn") || err.getMessage.contains("UNIXProcess"))) {
          return "(error executing: " + cmd + ")"
        }
        throw err
    } finally {
      if (process != null) {
        IOUtils.closeQuietly(process.getOutputStream)
        IOUtils.closeQuietly(process.getInputStream)
        IOUtils.closeQuietly(process.getErrorStream)
      }
    }
  }
}

File: gear-commons/src/main/java/cn/howardliu/gear/commons/utils/SystemInfo.java
package cn.howardliu.gear.commons.utils;

import java.lang.management.ManagementFactory;
import java.lang.management.OperatingSystemMXBean;
import java.lang.management.RuntimeMXBean;
import java.lang.reflect.Method;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

/**
 * <br>created at 2020/5/5
 *
 * @author liuxh
 * @since 1.0.0
 */
public class SystemInfo {
    private final static double KB = 1024;
    private final static double MB = 1024 * KB;
    private final static double GB = 1024 * MB;
    private final static DecimalFormat decimalFormat = new DecimalFormat("#.##", DecimalFormatSymbols.getInstance(Locale.ROOT));

    public static Map<String, Object> getSystemInfo() {
        final Map<String, Object> info = new HashMap<>();

        final OperatingSystemMXBean os = ManagementFactory.getOperatingSystemMXBean();

        info.put("name", os.getName());
        info.put("version", os.getVersion());
        info.put("arch", os.getArch());
        info.put("systemLoadAverage", os.getSystemLoadAverage());

        // com.sun.management.OperatingSystemMXBean
        info.put("committedVirtualMemorySize", addGetterIfAvaliable(os, "committedVirtualMemorySize"));
        info.put("freePhysicalMemorySize", addGetterIfAvaliable(os, "freePhysicalMemorySize"));
        info.put("freeSwapSpaceSize", addGetterIfAvaliable(os, "freeSwapSpaceSize"));
        info.put("processCpuTime", addGetterIfAvaliable(os, "processCpuTime"));
        info.put("totalPhysicalMemorySize", addGetterIfAvaliable(os, "totalPhysicalMemorySize"));
        info.put("totalSwapSpaceSize", addGetterIfAvaliable(os, "totalSwapSpaceSize"));

        // com.sun.management.UnixOperatingSystemMXBean
        info.put("openFileDescriptorCount", addGetterIfAvaliable(os, "openFileDescriptorCount"));
        info.put("maxFileDescriptorCount", addGetterIfAvaliable(os, "maxFileDescriptorCount"));

        try {
            if (!os.getName().toLowerCase(Locale.ROOT).startsWith("windows")) {
                info.put("uname", Processor.execute("uname -a"));
                info.put("uptime", Processor.execute("uptime"));
            }
        } catch (Exception ignored) {
        }
        return info;
    }

    public static Map<String, Object> getJvmInfo() {
        final Map<String, Object> jvm = new HashMap<>();

        final String javaVersion = System.getProperty("java.specification.version", "unknown");
        final String javaVendor = System.getProperty("java.specification.vendor", "unknown");
        final String javaName = System.getProperty("java.specification.name", "unknown");
        final String jreVersion = System.getProperty("java.version", "unknown");
        final String jreVendor = System.getProperty("java.vendor", "unknown");
        final String vmVersion = System.getProperty("java.vm.version", "unknown");
        final String vmVendor = System.getProperty("java.vm.vendor", "unknown");
        final String vmName = System.getProperty("java.vm.name", "unknown");

        jvm.put("version", jreVersion + " " + vmVersion);
        jvm.put("name", jreVendor + " " + vmName);

        final Map<String, String> specMap = new HashMap<>();
        specMap.put("vendor", javaVendor);
        specMap.put("name", javaName);
        specMap.put("version", javaVersion);
        jvm.put("spec", specMap);

        final Map<String, String> jreMap = new HashMap<>();
        jreMap.put("vendor", jreVendor);
        jreMap.put("version", jreVersion);
        jvm.put("jre", jreMap);

        final Map<String, String> vmMap = new HashMap<>();
        vmMap.put("vendor", vmVendor);
        vmMap.put("name", vmName);
        vmMap.put("version", vmVersion);
        jvm.put("vm", vmMap);

        final Runtime runtime = Runtime.getRuntime();
        jvm.put("processors", runtime.availableProcessors());

        final long freeMemory = runtime.freeMemory();
        final long maxMemory = runtime.maxMemory();
        final long totalMemory = runtime.totalMemory();
        final long usedMemory = totalMemory - freeMemory;
        final double percentUsed = usedMemory * 100.0 / maxMemory;

        Map<String, Object> memMap = new HashMap<>();
        memMap.put("free", humanReadableUnits(freeMemory, decimalFormat));
        memMap.put("total", humanReadableUnits(totalMemory, decimalFormat));
        memMap.put("max", humanReadableUnits(maxMemory, decimalFormat));
        memMap.put("used", humanReadableUnits(usedMemory, decimalFormat) + "(" + decimalFormat.format(percentUsed) + "%)");

        Map<String, Number> raw = new HashMap<>();
        raw.put("free", freeMemory);
        raw.put("used", usedMemory);
        raw.put("used%", percentUsed);
        raw.put("total", totalMemory);
        raw.put("max", maxMemory);
        memMap.put("raw", raw);

        jvm.put("memory", memMap);

        final RuntimeMXBean mx = ManagementFactory.getRuntimeMXBean();

        Map<String, Object> jmx = new HashMap<>();
        jmx.put("bootclasspath", mx.getBootClassPath());
        jmx.put("classpath", mx.getClassPath());
        jmx.put("commandLineArgs", mx.getInputArguments());
        jmx.put("startTime", mx.getStartTime());
        jmx.put("upTimeMS", mx.getUptime());

        jvm.put("jmx", jmx);

        return jvm;
    }

    private static String humanReadableUnits(final long bytes, final DecimalFormat df) {
        final double gbBytes = bytes / GB;
        if ((long) gbBytes > 0) {
            return df.format(gbBytes) + "GB";
        }
        final double mbBytes = bytes / MB;
        if ((long) mbBytes > 0) {
            return df.format(mbBytes) + "MB";
        }
        final double kbBytes = bytes / KB;
        if ((long) kbBytes > 0) {
            return df.format(kbBytes) + "KB";
        }
        return df.format(bytes) + "B";
    }

    private static Object addGetterIfAvaliable(final Object obj, String getter) {
        try {
            final String n = Character.toUpperCase(getter.charAt(0)) + getter.substring(1);
            final Method m = obj.getClass().getMethod("get" + n);
            return m.invoke(obj);
        } catch (Exception ignored) {
            return "";
        }
    }
}
